### PR TITLE
Usability improvements for z_importkey

### LINKED ISSUE #2032: Allow users to specify a portion of the blockchain to scan when importing z_addresses
https://github.com/zcash/zcash/pull/1978#issuecomment-273937381

Bearing in mind that some users may know in advance that the z_addresses they're importing only have transactions from a particular point in the blockchain, give them the option to avoid needlessly scanning prior blocks.

### PR DESCRIPTION
**Pull request by str4d**
- Add height parameter to z_importkey to reduce rescan range
- Change semantics of rescan parameter, so users can explicitly force a rescan
  for existing keys.

Closes #2032

### CODE REVIEW

**Comment by str4d** [comment]
Currently the `z_importkey` RPC behaves correctly with the new parameter, but the RPC test fails because when restarting the node with `-rescan` the older transactions are not discovered.

**Comment by daira** [comment]
I'm assuming this should not be reviewed yet because it has [WIP].

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -601,7 +609,7 @@ Value z_importkey(const Array& params, bool fHelp)
 
         // We want to scan for transactions and notes
         if (fRescan) {
-            pwalletMain->ScanForWalletTransactions(chainActive.Genesis(), true);
+            pwalletMain->ScanForWalletTransactions(chainActive[nRescanHeight], true);
```
Comment (daira) [inline src/wallet/rpcdump.cpp]: What happens if ``nRescanHeight`` is too large?

**Comment by str4d** [comment]
CChain::operator[] returns NULL for invalid heights, and
ScanForWalletTransactions is a no-op for NULL.

On 9 Feb 2017 8:36 pm, "Daira Hopwood" <notifications@github.com> wrote:

*@daira* requested changes on this pull request.
------------------------------

In src/wallet/rpcdump.cpp
<https://github.com/zcash/zcash/pull/2052#pullrequestreview-21112800>:

> @@ -601,7 +609,7 @@ Value z_importkey(const Array& params, bool fHelp)

         // We want to scan for transactions and notes
         if (fRescan) {
-            pwalletMain->ScanForWalletTransactions(chainActive.Genesis(),
true);
+            pwalletMain->ScanForWalletTransactions(chainActive[nRescanHeight],
true);

What happens if nRescanHeight is too large?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/zcash/zcash/pull/2052#pullrequestreview-21112800>, or mute
the thread

**Comment by str4d** [comment]
@bitcartel I have not had time to figure out why the RPC test is failing, so this should be bumped from 1.0.6.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (str4d) [inline src/wallet/rpcdump.cpp]: `CChain::operator[]` returns `NULL` for invalid heights, and `ScanForWalletTransactions` is a no-op for `NULL`.

Comment (bitcartel) [inline src/wallet/rpcdump.cpp]: It's not really a no-op per se.  There are side effects such as the wallet getting locked and ShowProgress getting invoked.  Would be safer for z_importkey to return an RPC error if out of range and ScanForWalletTransactions to immediately check for null when invoked.

**Comment by bitcartel** [comment]
I added a comment above about nRescanHeight.

**Comment by arcalinea** [comment]
If we can figure out why the RPC test is failing and fix it by tomorrow, this will make it into 1.0.7. Otherwise I'll push it into 1.0.8

**Comment by arcalinea** [comment]
The fix in #1978 does a rescan when a new key is imported with z_importkey:

>  // whenever a key is imported, we need to scan the whole chain
     pwalletMain->nTimeFirstKey = 1; // 0 would be considered 'no value'

But in this new test case we're only scanning part of the chain when the key is initially imported, so when we `-rescan`, won't the behavior default back to what was originally reported as a problem in #1941 with z_importkey?

> However, the rescan code skips over blocks that are older than the time the wallet was created, so if the node is a new installation, it will not see historical transactions.

**Comment by str4d** [comment]
It shouldn't skip any blocks, because when adding a key manually, the key creation time is set to 1 (because we can't know when it was actually created). But there's clearly an inconsistency between what we think should be happening, and what the test is trying and failing to do.

**Comment by arcalinea** [comment]
Ok, seems to be some issue with timing of when rescan starts from. Bumping this out

**Comment by str4d** [comment]
Added logging at the start of `ScanForWalletTransactions`, and sure enough:

```
  nTimeFirstKey: 1
  nTimeFirstKey: 1388543400
```

So after node 1 is stopped and restarted, the `-rescan` is not correctly picking up older blocks because it thinks the oldest key is after node 0 made its transaction. Likely it is picking up the creation time of an internal address. We could create node 1 earlier, but that would alter the original intention of the test I am reusing.

This inherently feels like a bug in `-rescan`, but I'm not sure how we could ever reliably resolve this. The whole point of the "earliest key creation time" mechanism is to avoid rescanning the entire block chain, and this PR is trying to extend that to adding z-addrs.

I am personally more in favour of adjusting the semantics of `z_importkey` when it is called twice with the same key:

- On first call (ie. if the key in question is not in the wallet), `rescan` defaults to true; for subsequent calls, `rescan` defaults to false.
- If `rescan=true`, always scan the blockchain even if we already have the key.

For the default settings, the current API behaviour is maintained (duplicate imports are silently ignored and don't trigger a rescan). But if a user imports a key scanning from some height `X`, and later wants to rescan from height `Y` (which could be any height, not just less than `X`), they can issue `z_importkey key true Y`.

**Comment by daira** [comment]
If @str4d's approach is adopted, I suggest having a way to spell the default semantics explicitly, e.g. rescan = ``whenkeyisnew``. Apart from the general principle of Explicit Is Better Than Implicit, this allows you to say ``z_importkey $KEY whenkeyisnew $HEIGHT``, which would mean "rescan from $HEIGHT iff the key is not already in the wallet".

**Comment by str4d** [comment]
@daira implemented my approach with your suggestion.

**Comment by daira** [comment]
utACK

**Comment by bitcartel** [comment]
ACK but would like:
- clearer documentation in help about parameter as I think it will confuse users with:` boolean or "whenkeyisnew"`
- for examples, `"whenkeyisnew"` doesn't have to be in quotes, when used on actual command line
- perhaps simplify and make the argument a string: `2. rescan ("yes", "no", "whenkeyisnew")` and for backwards compatibility with any existing usage/scripts, convert boolean to yes/no

**Comment by str4d** [comment]
@bitcartel comments addressed. @daira you may want to review the new commit.

**Comment by daira** [comment]
The compatibility fallback to accepting a boolean is not tested.

**Comment by str4d** [comment]
@daira done.

**Comment by daira** [comment]
utACK 53c96bfa6ce96d7815774ee40d94024002b6a1e2. Needs another ACK.

**Comment by bitcartel** [comment]
ACK but when you supply a height<0 or height>tip, there is no error.  Nothing bad happens because the [] operator handles invalid range, `chainActive[nRescanHeight]`.

@daira I think we can merge and file a ticket about reporting an error in future, taking into account overall RPC design for height values: https://github.com/zcash/zcash/issues/2197

**Comment by str4d** [comment]
@bitcartel given that #2187 also contains a height parameter and gives an error, it makes sense that this RPC should do so as well for now (so that we are expanding valid parameters with #2197 rather than re-defining them). I'll ad it now.

**Comment by bitcartel** [comment]
ACK

**Comment by bitcartel** [comment]
@zkbot r+

**Comment by daira** [comment]
Post-hoc utACK 33589401ba7df5495a2e57c6c4edd113dff3bd8b

**Pull request merged**